### ISSUE TITLE
[libtorch]Same model in CUDA and CPU got different result?

### ISSUE THREAD

**Issue by kaonick** [pytorch/pytorch#26456]
## 🐛 Bug

I have already asked questions in the forum, the link is as follows,
https://discuss.pytorch.org/t/why-same-model-in-cuda-and-cpu-got-different-result/56241

I happened in the same program, using the same model, switching between the GPU and the CPU, and the results were different.

My libtorch version is libtorch1.2.0_GPU.
## To Reproduce

1.load model:

```
	torch::NoGradGuard no_grad_guard;
	if (isCUDA) {
		module = torch::jit::load(model_path, torch::kCUDA);
	}
	else {
		module = torch::jit::load(model_path);
	}
```

2.forward:
```
	if (isCUDA) {
		output = module.forward({ tensor_image.to(torch::kCUDA)});
	}
	else {
		output = module.forward({ tensor_image });
	}
```
3.treat output:
```
	}else if (output.isTuple()) {
		cout << "output isTuple!" << '\n';
		auto ouputTuple=output.toTuple();

		cout << "tuple_size:" << ouputTuple.use_count()<< '\n';
		torch::Tensor out_tensor = ouputTuple->elements()[0].toTensor();
		torch::Tensor out_tensor1 = ouputTuple->elements()[1].toTensor();
		cout << "out_tensor:" << out_tensor.sizes() << '\n';//out_tensor:[1, 288, 384, 2]  576,768
		cout << "out_tensor1:" << out_tensor1.sizes() << '\n';//out_tensor1:[1, 32, 288, 384]

		torch::Tensor score;
		if (isCUDA) {

			torch::Tensor new_out_tensor = out_tensor.cpu();// .to(torch::kCPU).detach();
			score = new_out_tensor.squeeze(0);// [288, 384, 2]
		}
		else {
			score = out_tensor.squeeze(0);// [288, 384, 2]
		}
```

## Expected behavior

GPU and CPU should have the same output.

## Environment

Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Microsoft Windows 10 專業版
GCC version: Could not collect
CMake version: version 3.14.4

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 426.00
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.4+mkl
[pip3] numpydoc==0.8.0
[pip3] torch==1.2.0
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.4.0
[pip3] torchviz==0.0.1
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.1                      144
[conda] mkl-service               1.1.2            py36hb782905_5
[conda] mkl_fft                   1.0.6            py36h6288b17_0
[conda] mkl_random                1.0.2            py36h343c172_0
[conda] numpy                     1.16.4+mkl               pypi_0    pypi
[conda] torch                     1.1.0                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.4.0                    pypi_0    pypi
[conda] torchviz                  0.0.1                    pypi_0    pypi

**Comment by mrshenli** [comment]
@kaonick does the problem still occur if you build the model in the code instead of loading it? It will be helpful if you could share a minimum repro, including input data.

**Comment by mrshenli** [comment]
Based on the discussion in the [forum](https://discuss.pytorch.org/t/why-same-model-in-cuda-and-cpu-got-different-result/56241), using CPU gives the correct result, but not CUDA.

**Comment by kaonick** [comment]
@mrshenli I just use pretrain model(CRAFT) . so I can not build the model.
Or your mean is build the model then load weights?

**Comment by kaonick** [comment]
I put my model and code on github, if you can, please try it out (please pay attention to the operation instructions, choose 1 for CUDA, 2 for CPU).
Https://github.com/kaonick/libtorch-gpu
You should know it, it seems very strange.


Because the model is not small, please note that if you use the GPU, the GPU RAM should not be too small, otherwise it will be destroyed.

**Comment by kaonick** [comment]
Could anyone help this...

**Comment by FantasyJXF** [comment]
> @mrshenli I just use pretrain model(CRAFT) . so I can not build the model.
> Or your mean is build the model then load weights?

Same here, I traced CRAFT model, and the results between CPU and CUDA are different, that's really strange, the Python code with CPU/CUDA is all right.

Have you solved the bug?

**Comment by kaonick** [comment]
> > @mrshenli I just use pretrain model(CRAFT) . so I can not build the model.
> > Or your mean is build the model then load weights?
> 
> Same here, I traced CRAFT model, and the results between CPU and CUDA are different, that's really strange, the Python code with CPU/CUDA is all right.
> 
> Have you solved the bug?

No.Not yet.
CRAFT is really good,So I deploy in python to run then call by socket directly.
Speed is fast then run in Libtorch.

**Comment by FantasyJXF** [comment]
@kaonick  Flask + gunicorn + pytorch for my case.

**Comment by FantasyJXF** [comment]
Hi, @kaonick . I have solved the problem.

The problem comes from moving `torch::Tentor` data to `cv::Mat` with the following line:
```
std::memcpy((void *) textmap.data, score_text.data_ptr(), torch::elementSize(torch::kFloat) * score_text.numel());
```

The reason is supposed to be `contiguous` operation within pytorch, the libtorch-cuda maybe has some bug moving data from CUDA to CPU, we could check the model outputs from CPU/CUDA device, they are the same. But if you move tensor data from CPU to `cv::Mat` with memory copy, it would cause data NOT CONTINUOUS.

As for CRAFT model, just remove the permute operation when you trace the model.

As is known, `permute`, `view`, `narrow()` , `expand()` and `transpose()` suffers from contiguous.

Anyway, I solve the problem by doing this, NOT SURE whether it’s the hidden reason.

Hope this would help.

**Comment by kaonick** [comment]
> The reason is supposed to be `contiguous` operation within pytorch, the libtorch-cuda maybe has some bug moving data from CUDA to CPU, we could check the model outputs from CPU/CUDA device, they are the same. But if you move tensor data from CPU to `cv::Mat` with memory copy, it would cause data NOT CONTINUOUS.
> 
> As for CRAFT model, just remove the permute operation when you trace the model.
> 
> As is known, `permute`, `view`, `narrow()` , `expand()` and `transpose()` suffers from contiguous.
> 
> Anyway, I solve the problem by doing this, NOT SURE whether it’s the hidden reason.
> 
> Hope this would help.

OW~~~you so good.
That's too hard to me....
Your mean is when traced_script_module pth to pt just use random input not any preprocessing(like permute...) ?
But why ?
Why will cause this....

**Comment by khahn0213** [comment]
Hello, @FantasyJXF. May I ask you whether you don't use 
`permute`, `view`, `narrow()` , `expand()` and `transpose()` 
or you add `.contiguous()` at the end?
My code has lots of them, so I'm looking forward to your reply :-)

**Comment by FantasyJXF** [comment]
@khahn0213  I didn't add `.contiguous`, but I think add this might solve the problem.

**Comment by khahn0213** [comment]
@FantasyJXF Thanks for your reply. So you just remove those pytorch functions and write the code in another way??

**Comment by FantasyJXF** [comment]
Yes, it work fine for me

---
From Fantasy

**Comment by rlczddl** [comment]
I have this error, too.And I use random input without any preprocessing when tracing my model.The problem comes from moving torch::Tentor data to Eigen::MatriXf with the following line:
    auto tt = t.contiguous().to(torch::kCPU).contiguous(); // t is at::Tensor and t.dim()==2
    Eigen::Map<MatrixType<float, Eigen::RowMajor>> e(tt.data_ptr<float>(), t.size(0), t.size(1));

Any help?

**Comment by rlczddl** [comment]
Hi, I think I found the reason.In my case, i  wrote a conversion function:
auto tensor2eigen(at::Tensor& t)
{
    assert(t.dim()==2);
    auto tt = t.to(torch::kCPU);
    Eigen::Map<MatrixType<float, Eigen::RowMajor>> e(tt.data_ptr<float>(), t.size(0), t.size(1));
    return e;
}
When i got an output from network(output on torch::kCUDA), i put it to this function,  and outliers will appear in the result matrix.But if i first transform output to torch::kCPU, then put it to this function, and finally  i got correct result.

**Comment by ghgggg** [comment]
the main reason is that  libtorch op to(torch::kCPU) is asynchronous；
so we need this firstly
tensor = tensor.to(torch::kCPU); 
then do other ops.
It should work
